mantidproject/mantid
# Waterfall Plots in Workbench

**Description of work.**
This PR adds waterfall plots to Workbench. These are a type of plot where each line is offset from the previous one by some amount.

![image](https://user-images.githubusercontent.com/52415735/70530292-799bb980-1b4a-11ea-9685-ce66f4c79a6a.png)

**To test:**
Here is a list of things to check (have fun):
- In the spectra selector dialog which opens when you double-click a workspace, in the plot type drop-down list there should now be a Waterfall option.
- Choose Waterfall and enter multiple spectra and a plot with the curves spaced apart should be produced.
- There should be three new options in the toolbar:
    - Change offset: this should open a dialog allowing you to change the amount by which each curve is separated in the x and y direction. There is also a check box which keeps the proportion that the x and y values are currently in, e.g. if you double one the other will also double.
    - Reverse the order of the curves on the plot.
    - Fill the area under each curve: this should open a dialog allowing you to enable or disable the area under each curve being filled in, and also set whether the fills are the same colour as each line or a single custom colour.
- The toolbar should not have the button to generate a script when the plot is a waterfall plot because plot-to-script currently doesn't work for waterfall plots (an issue will be made for this).
- When you right-click the plot there should now be a Plot Type option where you can switch between a regular 1D plot and a waterfall plot. Changing from Waterfall to 1D should remove the filled areas if they have been enabled, return each curve to its original position, remove the waterfall options from the toolbar, and restore the generate script option.
- Changing any of the figure options with a waterfall plot should maintain the waterfall-ness.
- If the fill under each curve has been enabled and is set to be the same colour as each line, changing the colour of a line in the figure options should change the colour of its fill.
- Removing a curve via the figure options should redo the waterfall so that the spacing between each curve is equal.
- When you overplot onto a waterfall plot the new curves should "join" the waterfall. If the existing curves were filled, then the new curves should also be filled with the correct colour.
- You should not be able to create a waterfall plot with only one curve. This means that in the spectra selector dialog, if you choose Waterfall as the plot type but only enter one spectrum number, a regular 1D plot should be produced. There should be no option to change the plot type if you right-click a plot with only one curve. If you plot multiple curves as a waterfall plot and then remove every curve but one using the figure options, the plot should become a regular 1D plot. You can tell this is the case because the waterfall options will no longer be visible in the toolbar and if the curve was filled, the fill should be removed.
- When you right-click a tiled plot, there should be no option to change the plot type because you can't have a weird waterfall-tiled hybrid plot.

Refs #27450 

No release notes because it's probably a good idea to wait until all the new plotting stuff has been added to Workbench to decide how the release notes should be.

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.
